Realm/ObjectStore/object_store_exceptions.hpp

```diff
@@ -24,42 +24,78 @@
 #include <string>
 
 namespace realm {
+    class Property;
 
     class ObjectStoreException : public std::exception {
     public:
         enum class Kind {
-            // thrown when calling update_realm_to_schema and the realm version is greater than the given version
-            RealmVersionGreaterThanSchemaVersion,   // old_version, new_version
-            RealmPropertyTypeNotIndexable,          // object_type, property_name, property_type
-            RealmDuplicatePrimaryKeyValue,          // object_type, property_name
+            RealmVersionGreaterThanSchemaVersion,   // OldVersion, NewVersion
+            RealmPropertyTypeNotIndexable,          // ObjectType, PropertyName, PropertyType
+            RealmDuplicatePrimaryKeyValue,          // ObjectType, PropertyName, PropertyType
+            ObjectSchemaMissingProperty,            // ObjectType, PropertyName, PropertyType
+            ObjectSchemaNewProperty,                // ObjectType, PropertyName, PropertyType
+            ObjectSchemaMismatchedTypes,            // ObjectType, PropertyName, PropertyType, OldPropertyType
+            ObjectSchemaMismatchedObjectTypes,      // ObjectType, PropertyName, PropertyType, ObjectType, OldObjectType
+            ObjectSchemaChangedPrimaryKey,          // ObjectType, PrimaryKey
+            ObjectSchemaNewPrimaryKey,              // ObjectType, PrimaryKey
+            ObjectSchemaChangedOptionalProperty,
+            ObjectSchemaNewOptionalProperty,
+            ObjectStoreValidationFailure,           // ObjectType, vector<ObjectStoreException>
         };
-        typedef std::map<std::string, std::string> Dict;
 
-        ObjectStoreException(Kind kind, Dict dict = Dict());
+        typedef const std::string InfoKey;
+        typedef std::map<InfoKey, std::string> Info;
```

Comment:
Consider using `using` instead of `typedef` (easier to grep for).

```c++
using InfoKey = const std::string;
using Info = std::map<InfoKey, std::string>;
```

Also note that the "key" template parameter passed to `std::map` is turned `const` anyway, so unless you explicitly want to forbid modifications to an `InfoKey` *before* it gets to the map, it could be just `std::string` without the `const`.